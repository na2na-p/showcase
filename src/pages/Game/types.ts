export type historyType = {
	index: number;
	history: string;
	result: string;
}
