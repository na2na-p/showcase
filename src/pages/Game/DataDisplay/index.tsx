import React from 'react';

import Table from '@/components/dataDisplay/Table';

import type { historyType } from '../types';

export default ({ histories }: { histories: historyType[] }) => {
	const tableHeaders = ['回数', 'あなたの答え', '判定'];
	const historyRows = histories.map((historyRecord, index) => {
		const { history, result } = historyRecord;
		return [index + 1, history, result];
	});
	return (
		<Table headers={tableHeaders} rows={historyRows} />
	);
};
