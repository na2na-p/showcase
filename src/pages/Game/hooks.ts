import { useState } from 'react';

import type { HistoryType } from './types';

export const useHooks = ({
	answer
}: {
	answer?: number
}) => {
	const initialAnswer = answer || Math.floor(Math.random() * 100) + 1;
	const [histories, setHistories] = useState<HistoryType[]>([]);
	const [answerState] = useState(initialAnswer);

	const setResultHistory = (result: string) => {
		setHistories([...histories, {
			index: histories.length + 1,
			history: result,
			// answerとresultを比較して、大きいか小さいか正解かを判定する
			result: parseInt(result) === answerState ?
				'正解' : parseInt(result) > answerState ?
					'もっと小さい' : 'もっと大きい'
		}]);
	};

	const onResetButtonClick = () => {
		setHistories([]);
	};

	return {
		histories,
		setResultHistory,
		onResetButtonClick
	};
};
