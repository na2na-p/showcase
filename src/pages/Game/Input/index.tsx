import React from 'react';

import TextField from '@/components/input/TextField';
import Button from '@/components/input/Button';

import { useHooks } from './hooks';

const Input = ({ setter }: { setter: (result: string) => void }) => {
	const { inputRef, onButtonClick } = useHooks({ setter });
	return (
		<>
			<TextField placeholder='いくつでしょう?' inputRef={inputRef} type="number" />
			<Button onClick={onButtonClick} variant="contained">トライ!</Button>
		</>
	);
};

export default Input;
