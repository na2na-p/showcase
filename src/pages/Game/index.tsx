import React from 'react';

import Button from '@/components/input/Button';
import Stack from '@/components/layout/Stack';
import Divider from '@/components/layout/Divider';
import Typography from '@/components/layout/Typography';

import Input from './Input';
import DataDisplay from './DataDisplay';
import { useHooks } from './hooks';

export type GameProps = {
	answer?: number;
}

const Game = ({ answer }: GameProps) => {
	const { histories, setResultHistory, onResetButtonClick } = useHooks({ answer });
	return (
		<Stack spacing={2}>
			<Typography variant="h5">数当てゲーム</Typography>
			<Divider />
			<Input setter={setResultHistory} />
			<Divider />
			<DataDisplay histories={histories} />
			<Button onClick={onResetButtonClick} variant="contained">もう一度最初から</Button>
		</Stack>
	);
};

export default Game;
