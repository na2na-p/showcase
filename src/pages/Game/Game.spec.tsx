// import '@testing-library/jest-dom'; // jestのアサーションがエラーになる場合は明示的にimport
import { describe, expect, test } from 'vitest';
import Game from '.';
import { render, screen } from '@testing-library/react';
import { act } from 'react-dom/test-utils';

describe('Game component', () => {
	test('レンダリングテスト', () => {
		const { container } = render(<Game />);
		expect(container).toMatchInlineSnapshot();
	});

	describe('数当てテスト', () => {
		test('正解した場合', () => {
			render(<Game answer={20} />);
			const input: HTMLInputElement = screen.getByPlaceholderText('いくつでしょう?');
			const button: HTMLButtonElement = screen.getByRole('button', { name: 'トライ!' });
			act(() => {
				input.value = '20';
				button.click();
			});
			expect(screen.getByText('正解')).toBeInTheDocument();
		});

		test('数が大きすぎた場合', () => {
			render(<Game answer={20} />);
			const input: HTMLInputElement = screen.getByPlaceholderText('いくつでしょう?');
			const button: HTMLButtonElement = screen.getByRole('button', { name: 'トライ!' });
			act(() => {
				input.value = '30';
				button.click();
			});
			expect(screen.getByText('もっと小さい')).toBeInTheDocument();
		});

		test('数が小さすぎた場合', () => {
			render(<Game answer={20} />);
			const input: HTMLInputElement = screen.getByPlaceholderText('いくつでしょう?');
			const button: HTMLButtonElement = screen.getByRole('button', { name: 'トライ!' });
			act(() => {
				input.value = '10';
				button.click();
			});
			expect(screen.getByText('もっと大きい')).toBeInTheDocument();
		});
	});
});
