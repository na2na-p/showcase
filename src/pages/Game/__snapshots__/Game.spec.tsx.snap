// Vitest Snapshot v1

exports[`Game component > レンダリングテスト 1`] = `
<div>
  <div
    class="css-1p5q5e5-MuiStack-root"
  >
    <h4
      class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
    >
      数当てゲーム
    </h4>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-ctpjxq-MuiDivider-root"
    />
    <div
      class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    >
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          defaultvalue=""
          id=":r0:"
          placeholder="いくつでしょう?"
          type="number"
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-ihdtdm"
          >
            <span
              class="notranslate"
            >
              ​
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      トライ!
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-ctpjxq-MuiDivider-root"
    />
    <div
      class="MuiTableContainer-root css-rorn0c-MuiTableContainer-root"
    >
      <table
        class="MuiTable-root css-rqglhn-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              回数
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              あなたの答え
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              判定
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        />
      </table>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      もう一度最初から
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
